File: donater-bot/handlers/order.js
import { Composer } from 'grammy';
import { randomInt } from 'crypto';
import { productById, gamesListActive } from '../database/db.js';
import {
  isCaptchaBlocked,
  registerCaptchaFail,
  resetCaptchaFails,
  orderRateLimitMinutes,
} from '../middlewares/antiSpam.js';
import {
  captchaKeyboard,
  confirmOrderKeyboard,
  cancelInline,
  paymentMethodsKeyboard,
  mainMenuKeyboardJson,
} from '../utils/keyboards.js';
import { sendMessageRaw } from '../utils/tg_api.js';

export const orderComposer = new Composer();

function shuffle(arr) {
  const a = [...arr];
  for (let i = a.length - 1; i > 0; i--) {
    const j = randomInt(0, i + 1);
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}

orderComposer.callbackQuery(/^prod:(\d+)$/, async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  const chatId = ctx.chat?.id;
  if (!chatId) return;

  try {
    const uid = ctx.from?.id;
    if (!uid) return;

    const blockMin = isCaptchaBlocked(uid);
    if (blockMin) {
      await ctx.reply(
        ctx.i18n('captcha_blocked', { minutes: blockMin }),
        { parse_mode: 'HTML' }
      );
      return;
    }

    const rateMin = orderRateLimitMinutes(uid);
    if (rateMin > 0) {
      await ctx.reply(
        ctx.i18n('rate_limit_orders', { minutes: rateMin }),
        { parse_mode: 'HTML' }
      );
      return;
    }

    const pid = parseInt(ctx.match[1], 10);
    const p = productById(pid);
    if (!p) {
      await ctx.reply(ctx.i18n('error_generic'));
      return;
    }

    const gameId = ctx.session.selectedGameId;
    const gameName = ctx.session.selectedGameName || '';
    if (!gameId || p.game_id !== gameId) {
      await ctx.reply(ctx.i18n('error_generic'));
      return;
    }

    ctx.session.draft = {
      gameId,
      gameName,
      productId: pid,
      productLabel: p.label,
      priceTjs: p.price_tjs,
    };

    const a = randomInt(1, 9);
    const b = randomInt(1, 9);
    const correct = a + b;
    const wrong = new Set();
    while (wrong.size < 3) {
      const x = randomInt(2, 18);
      if (x !== correct) wrong.add(x);
    }
    const opts = shuffle([
      { label: correct, value: correct, correct: true },
      ...[...wrong].map((v) => ({ label: v, value: v, correct: false })),
    ]);
    const salt = `${randomInt(1000, 9999)}`;
    ctx.session.captchaSalt = salt;
    ctx.session.captchaAnswer = correct;
    ctx.session.step = 'captcha';

    const q = `${a} + ${b}`;
    await sendMessageRaw(chatId, ctx.i18n('captcha_title', { question: q }), {
      parse_mode: 'HTML',
      reply_markup: captchaKeyboard(salt, opts),
    });
  } catch (e) {
    console.error(e);
    await ctx.reply(ctx.i18n('error_generic'));
  }
});

orderComposer.callbackQuery(/^cap:(\w+):(\d+)$/, async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  try {
    const uid = ctx.from?.id;
    const salt = ctx.match[1];
    const val = parseInt(ctx.match[2], 10);
    if (ctx.session.step !== 'captcha' || salt !== ctx.session.captchaSalt) {
      return;
    }
    if (val === ctx.session.captchaAnswer) {
      resetCaptchaFails(uid);
      ctx.session.step = 'await_game_id';
      ctx.session.captchaSalt = null;
      ctx.session.captchaAnswer = null;
      await ctx.reply(ctx.i18n('enter_game_id'), {
        parse_mode: 'HTML',
        reply_markup: cancelInline(ctx.i18n),
      });
      return;
    }
    const rec = registerCaptchaFail(uid);
    const left = 3 - (rec.fails || 0);
    if (rec.blockedUntil && Date.now() < rec.blockedUntil) {
      await ctx.reply(
        ctx.i18n('captcha_blocked', {
          minutes: Math.ceil((rec.blockedUntil - Date.now()) / 60000) || 5,
        }),
        { parse_mode: 'HTML' }
      );
      return;
    }
    await ctx.reply(ctx.i18n('captcha_wrong', { n: Math.max(0, left) }), {
      parse_mode: 'HTML',
    });
  } catch (e) {
    console.error(e);
  }
});

orderComposer.callbackQuery('flow:cancel', async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  try {
    ctx.session.flow = null;
    ctx.session.step = null;
    ctx.session.draft = null;
    const chatId = ctx.chat?.id;
    if (!chatId) return;
    await sendMessageRaw(chatId, ctx.i18n('order_cancelled'), {
      parse_mode: 'HTML',
      reply_markup: mainMenuKeyboardJson(gamesListActive(), ctx.i18n),
    });
  } catch (e) {
    console.error(e);
  }
});

orderComposer.callbackQuery('order:confirm_no', async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  try {
    ctx.session.step = null;
    ctx.session.draft = null;
    const chatId = ctx.chat?.id;
    if (!chatId) return;
    await sendMessageRaw(chatId, ctx.i18n('order_cancelled'), {
      parse_mode: 'HTML',
      reply_markup: mainMenuKeyboardJson(gamesListActive(), ctx.i18n),
    });
  } catch (e) {
    console.error(e);
  }
});

orderComposer.callbackQuery('order:confirm_yes', async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  try {
    if (ctx.session.step !== 'confirm') return;
    ctx.session.step = 'choose_payment';
    const chatId = ctx.chat?.id;
    if (!chatId) return;
    await sendMessageRaw(chatId, ctx.i18n('choose_payment'), {
      parse_mode: 'HTML',
      reply_markup: paymentMethodsKeyboard(ctx.i18n),
    });
  } catch (e) {
    console.error(e);
  }
});

orderComposer.callbackQuery('flow:back_summary', async (ctx) => {
  await ctx.answerCallbackQuery().catch(() => {});
  try {
    const d = ctx.session.draft;
    if (!d || !d.accountId) return;
    ctx.session.step = 'confirm';
    const un = ctx.from?.username ? `@${ctx.from.username}` : '—';
    const summary = ctx.i18n('order_summary', {
      firstName: ctx.from?.first_name || '—',
      username: un,
      game: d.gameName,
      product: d.productLabel,
      price: d.priceTjs,
      accountId: d.accountId,
    });
    const chatId = ctx.chat?.id;
    if (!chatId) return;
    await sendMessageRaw(chatId, summary, {
      parse_mode: 'HTML',
      reply_markup: confirmOrderKeyboard(ctx.i18n),
    });
  } catch (e) {
    console.error(e);
  }
});

orderComposer.on('message:text', async (ctx, next) => {
  try {
    const text = ctx.message?.text?.trim() || '';
    if (text.startsWith('/')) return next();

    if (ctx.session.step !== 'await_game_id' || !ctx.session.draft) {
      return next();
    }

    if (!/^\d{8,12}$/.test(text)) {
      await ctx.reply(ctx.i18n('invalid_id'), { parse_mode: 'HTML' });
      return;
    }

    ctx.session.draft.accountId = text;
    ctx.session.step = 'confirm';

    const d = ctx.session.draft;
    const un = ctx.from?.username ? `@${ctx.from.username}` : '—';
    const summary = ctx.i18n('order_summary', {
      firstName: ctx.from?.first_name || '—',
      username: un,
      game: d.gameName,
      product: d.productLabel,
      price: d.priceTjs,
      accountId: d.accountId,
    });
    const chatId = ctx.chat?.id;
    if (!chatId) return;
    await sendMessageRaw(chatId, summary, {
      parse_mode: 'HTML',
      reply_markup: confirmOrderKeyboard(ctx.i18n),
    });
  } catch (e) {
    console.error(e);
    await ctx.reply(ctx.i18n('error_generic'));
  }
});

File: donater-bot/handlers/order.py
import logging
import random
import re
import time
from typing import Dict, List

from aiogram import F, Router
from aiogram.filters import StateFilter
from aiogram.fsm.context import FSMContext
from aiogram.types import CallbackQuery, InlineKeyboardButton, Message
from aiogram.utils.keyboard import InlineKeyboardBuilder

from database import queries
from handlers.start import _main_menu_markup
from middlewares.anti_spam import order_rate_limit_minutes
from states.order import OrderFlow
from utils import tg_api
from utils.formatter import escape_html

log = logging.getLogger(__name__)
router = Router(name="order")

_captcha_fails: Dict[int, Dict[str, float | int]] = {}
_CAPTCHA_BLOCK_MIN = 5


def _shuffle(opts: List[dict]) -> List[dict]:
    a = list(opts)
    random.shuffle(a)
    return a


@router.callback_query(F.data.regexp(re.compile(r"^prod:(\d+)$")))
async def cb_product(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    uid = query.from_user.id
    chat_id = query.message.chat.id
    try:
        block = _captcha_fails.get(uid, {}).get("blocked_until", 0)
        if isinstance(block, (int, float)) and time.monotonic() < float(block):
            m = max(1, int((float(block) - time.monotonic()) / 60) + 1)
            await query.message.answer(t("captcha_blocked", m=m), parse_mode="HTML")
            return

        wait = order_rate_limit_minutes(uid, minutes=2)
        if wait > 0:
            await query.message.answer(t("rate_limit", m=wait), parse_mode="HTML")
            return

        m = re.match(r"^prod:(\d+)$", query.data or "")
        pid = int(m.group(1)) if m else 0
        p = await queries.product_by_id(pid)
        if not p:
            await query.message.answer(t("error_generic"))
            return
        data = await state.get_data()
        gid = data.get("selected_game_id")
        if not gid or p["game_id"] != gid:
            await query.message.answer(t("error_generic"))
            return

        await state.update_data(
            draft={
                "game_id": gid,
                "game_name": data.get("selected_game_name", ""),
                "product_id": pid,
                "product_label": p["label"],
                "price_tjs": p["price_tjs"],
            },
            captcha_salt="",
            captcha_answer=0,
        )

        a, b = random.randint(1, 9), random.randint(1, 9)
        correct = a + b
        wrong = set()
        while len(wrong) < 3:
            x = random.randint(2, 18)
            if x != correct:
                wrong.add(x)
        opts = _shuffle(
            [{"label": correct, "value": correct, "correct": True}]
            + [{"label": v, "value": v, "correct": False} for v in wrong]
        )
        salt = str(random.randint(1000, 9999))
        await state.update_data(captcha_salt=salt, captcha_answer=correct)
        await state.set_state(OrderFlow.captcha)

        q = f"{a} + {b}"
        kb = tg_api.captcha_keyboard(salt, opts)
        await tg_api.raw_send_message(
            chat_id, t("captcha_title", q=q), reply_markup=kb, parse_mode="HTML"
        )
    except Exception as exc:
        log.exception("order product: %s", exc)
        await query.message.answer(t("error_generic"))


@router.callback_query(OrderFlow.captcha, F.data.regexp(re.compile(r"^cap:(\w+):(\d+)$")))
async def cb_captcha(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    uid = query.from_user.id
    m = re.match(r"^cap:(\w+):(\d+)$", query.data or "")
    if not m:
        return
    salt, val_s = m.group(1), m.group(2)
    val = int(val_s)
    data = await state.get_data()
    if data.get("captcha_salt") != salt:
        return
    correct = int(data.get("captcha_answer") or 0)
    if val == correct:
        _captcha_fails.pop(uid, None)
        await state.set_state(OrderFlow.game_id)
        await state.update_data(captcha_salt="", captcha_answer=0)
        b = InlineKeyboardBuilder()
        b.row(InlineKeyboardButton(text=t("confirm_no"), callback_data="flow:cancel"))
        await query.message.answer(
            t("enter_game_id"), parse_mode="HTML", reply_markup=b.as_markup()
        )
        return

    rec = _captcha_fails.setdefault(uid, {"fails": 0})
    rec["fails"] = int(rec.get("fails", 0)) + 1
    left = max(0, 3 - int(rec["fails"]))
    if rec["fails"] >= 5:
        rec["blocked_until"] = time.monotonic() + _CAPTCHA_BLOCK_MIN * 60
        await query.message.answer(
            t("captcha_blocked", m=_CAPTCHA_BLOCK_MIN), parse_mode="HTML"
        )
        return
    await query.message.answer(t("captcha_wrong", n=left), parse_mode="HTML")


@router.callback_query(F.data == "flow:cancel")
async def cb_flow_cancel(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    await state.clear()
    kb = await _main_menu_markup(t)
    await tg_api.raw_send_message(query.message.chat.id, t("order_cancelled"), reply_markup=kb)


@router.callback_query(F.data == "order:confirm_no")
async def cb_confirm_no(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    await state.clear()
    kb = await _main_menu_markup(t)
    await tg_api.raw_send_message(query.message.chat.id, t("order_cancelled"), reply_markup=kb)


@router.callback_query(OrderFlow.confirm, F.data == "order:confirm_yes")
async def cb_confirm_yes(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    await state.set_state(OrderFlow.payment_choice)
    kb = tg_api.payment_methods_markup(
        t("payment_dc"),
        t("payment_alif"),
        t("payment_mc"),
        t("payment_milli"),
        t("confirm_no"),
    )
    await tg_api.raw_send_message(
        query.message.chat.id, t("choose_payment"), reply_markup=kb, parse_mode="HTML"
    )


@router.callback_query(OrderFlow.confirm, F.data == "flow:back_summary")
async def cb_back_summary(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    data = await state.get_data()
    draft = data.get("draft") or {}
    if not draft.get("account_id"):
        return
    await state.set_state(OrderFlow.confirm)
    un = (
        f"@{query.from_user.username}"
        if query.from_user and query.from_user.username
        else "—"
    )
    txt = t(
        "order_summary",
        first_name=escape_html(query.from_user.first_name or "—"),
        username=escape_html(un),
        game=escape_html(draft.get("game_name", "")),
        product=escape_html(draft.get("product_label", "")),
        price=draft.get("price_tjs", 0),
        account_id=escape_html(draft.get("account_id", "")),
    )
    kb = tg_api.confirm_order_markup(t("confirm_yes"), t("confirm_no"))
    await tg_api.raw_send_message(query.message.chat.id, txt, reply_markup=kb, parse_mode="HTML")


@router.message(OrderFlow.game_id, F.text)
async def msg_game_id(message: Message, t, state: FSMContext):
    text = (message.text or "").strip()
    if text.startswith("/"):
        return
    if not re.fullmatch(r"\d{6,15}", text):
        await message.answer(t("invalid_id"), parse_mode="HTML")
        return
    data = await state.get_data()
    draft = dict(data.get("draft") or {})
    draft["account_id"] = text
    await state.update_data(draft=draft)
    await state.set_state(OrderFlow.confirm)
    un = f"@{message.from_user.username}" if message.from_user.username else "—"
    summary = t(
        "order_summary",
        first_name=escape_html(message.from_user.first_name or "—"),
        username=escape_html(un),
        game=escape_html(draft.get("game_name", "")),
        product=escape_html(draft.get("product_label", "")),
        price=draft.get("price_tjs", 0),
        account_id=escape_html(text),
    )
    kb = tg_api.confirm_order_markup(t("confirm_yes"), t("confirm_no"))
    await tg_api.raw_send_message(message.chat.id, summary, reply_markup=kb, parse_mode="HTML")
